**Gathering Data**

Pada tahap ini, kita akan mengumpulkan semua data yang dibutuhkan.

1. Memuat tabel customers

  Data pertama yang harus kita muat ialah data dari tabel customers.

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

customers_df = pd.read_csv("https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/main/DicodingCollection/customers.csv")
customers_df.head()

customer_id customer_name             gender  age  \
0            1       fulan 1             Female   30   
1            2       fulan 2  Prefer not to say   69   
2            3       fulan 3  Prefer not to say   59   
3            4       fulan 4  Prefer not to say   67   
4            5       fulan 5  Prefer not to say   30   

                     home_address  zip_code               city  \
0  8606 Victoria TerraceSuite 560      5464      Johnstonhaven   
1      8327 Kirlin SummitApt. 461      8223    New Zacharyfort   
2       269 Gemma SummitSuite 109      5661           Aliburgh   
3       743 Bailey GroveSuite 141      1729  South Justinhaven   
4         48 Hyatt ManorSuite 375      4032     Griffithsshire   

                          state    country  
0            Northern Territory  Australia  
1               South Australia  Australia  
2  Australian Capital Territory  Australia  
3                    Queensland  Australia  
4                    Queensland  Australia

2. Load tabel orders

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

orders_df = pd.read_csv("https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/main/DicodingCollection/orders.csv")
orders_df.head()

order_id  customer_id  payment order_date delivery_date
0         1           64    30811  2021-8-30    2021-09-24
1         2          473    50490   2021-2-3    2021-02-13
2         3          774    46763  2021-10-8    2021-11-03
3         4          433    39782   2021-5-6    2021-05-19
4         5          441    14719  2021-3-23    2021-03-24

3. Memuat tabel product

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

product_df = pd.read_csv("https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/main/DicodingCollection/products.csv")
product_df.head()

product_id product_type  product_name size colour  price  quantity  \
0           0        Shirt  Oxford Cloth   XS    red    114        66   
1           1        Shirt  Oxford Cloth    S    red    114        53   
2           2        Shirt  Oxford Cloth    M    red    114        54   
3           3        Shirt  Oxford Cloth    L    red    114        69   
4           4        Shirt  Oxford Cloth   XL    red    114        47   

                                    description  
0  A red coloured, XS sized, Oxford Cloth Shirt  
1   A red coloured, S sized, Oxford Cloth Shirt  
2   A red coloured, M sized, Oxford Cloth Shirt  
3   A red coloured, L sized, Oxford Cloth Shirt  
4  A red coloured, XL sized, Oxford Cloth Shirt

4. Load tabel sales

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

sales_df = pd.read_csv("https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/main/DicodingCollection/sales.csv")
sales_df.head()

sales_id  order_id  product_id  price_per_unit  quantity  total_price
0         0         1         218             106         2        212.0
1         1         1         481             118         1        118.0
2         2         1           2              96         3        288.0
3         3         1        1002             106         2        212.0
4         4         1         691             113         3        339.0

Assessing Data

 pada proses ini
kita akan menilai kualitas dari seluruh data yang akan digunakan. Penilaian
ini bertujuan untuk melihat berbagai permasalahan yang ada dalam data
tersebut.

1. Menilai Data customers_df

memeriksa tipe data data dari tiap kolom yang terdapat dalam
customers_df.

In [ ]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1007 non-null   int64 
 1   customer_name  1007 non-null   object
 2   gender         989 non-null    object
 3   age            1007 non-null   int64 
 4   home_address   1007 non-null   object
 5   zip_code       1007 non-null   int64 
 6   city           1007 non-null   object
 7   state          1007 non-null   object
 8   country        1007 non-null   object
dtypes: int64(3), object(6)
memory usage: 70.9+ KB


In [ ]:
customers_df.isna().sum()

customer_id       0
customer_name     0
gender           18
age               0
home_address      0
zip_code          0
city              0
state             0
country           0
dtype: int64

gambar di atas menunjukkan bahwa terdapat 18 missing values pada
kolom gender.

In [ ]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  6


menemukan bahwa terdapat 6
data yang terduplikat.

In [ ]:
customers_df.describe()

customer_id          age     zip_code
count  1007.000000  1007.000000  1007.000000
mean    501.726912    50.929494  5012.538232
std     288.673238    30.516299  2885.836112
min       1.000000    20.000000     2.000000
25%     252.500000    34.000000  2403.500000
50%     502.000000    50.000000  5087.000000
75%     751.500000    65.000000  7493.500000
max    1000.000000   700.000000  9998.000000

Sekarang kita memeriksa parameter statistik dari kolom numerik yang
terdapat dalam customers_df. Untuk mempermudah pemeriksaan, kita akan
menggunakan method describe(). Method tersebut akan menampilkan
ringkasan parameter statistik (mean, median, dll.) dari kolom numerik pada
sebuah DataFrame.

Jika diperhatikan, terdapat keanehan pada nilai maksimum yang terdapat
pada kolom age. Ini kemungkinan besar terjadi karena adanya inaccurate
value pada kolom tersebut.

2. Menilai Data orders_df

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   payment        1000 non-null   int64 
 3   order_date     1000 non-null   object
 4   delivery_date  1000 non-null   object
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


Jika Anda perhatikan, jumlah data pada orders_df telah lengkap atau dengan
kata lain tidak ada missing value di dalamnya. Akan tetapi, terdapat
kesalahan tipe data untuk kolom order_date & delivery_date. Kedua kolom
tersebut harusnya direpresentasikan sebagai tipe
data datetime, bukan object (ini digunakan untuk tipe data string).

In [ ]:
print("Jumlah duplikasi: ", orders_df.duplicated().sum())
orders_df.describe()

Jumlah duplikasi:  0


order_id  customer_id       payment
count  1000.000000  1000.000000   1000.000000
mean    500.500000   506.640000  33972.936000
std     288.819436   277.115502  14451.609047
min       1.000000     1.000000  10043.000000
25%     250.750000   275.250000  21329.250000
50%     500.500000   515.000000  33697.500000
75%     750.250000   737.250000  46249.000000
max    1000.000000  1000.000000  59910.000000

Jika diperhatikan, tidak ada keanehan pada hasil tersebut. Ini menunjukkan
tidak terdapat duplikasi dan keanehan nilai pada orders_df.  

3. Menilai Data product_df

gunakan
method info() untuk memastikan tipe data tiap kolom dalam product_df telah
sesuai.

In [ ]:
product_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1266 entries, 0 to 1265
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    1266 non-null   int64 
 1   product_type  1266 non-null   object
 2   product_name  1266 non-null   object
 3   size          1266 non-null   object
 4   colour        1266 non-null   object
 5   price         1266 non-null   int64 
 6   quantity      1266 non-null   int64 
 7   description   1266 non-null   object
dtypes: int64(3), object(5)
memory usage: 79.2+ KB


Berdasarkan hasil tersebut, dapat disimpulkan bahwa tidak terdapat masalah
pada tipe data tiap kolom dalam product_df.  Selain itu, jumlah datanya juga
telah lengkap sehingga tidak ada missing value di dalamnya.
Hal berikutnya yang harus diperiksa ialah duplikasi dan ringkasan parameter
statistik dari product_df. Berikut merupakan kode untuk melakukannya.

In [ ]:
print("Jumlah duplikasi: ", product_df.duplicated().sum())
product_df.describe()

Jumlah duplikasi:  6


product_id        price     quantity
count  1266.000000  1266.000000  1266.000000
mean    627.926540   105.812006    60.138231
std     363.971586     9.715611    11.682791
min       0.000000    90.000000    40.000000
25%     313.250000    95.250000    50.000000
50%     626.500000   109.000000    60.000000
75%     942.750000   114.000000    70.000000
max    1259.000000   119.000000    80.000000

Berdasarkan gambar di atas, dapat dilihat bahwa terdapat 6 data yang
terduplikasi pada product_df.

4. Menilai Data Sales_df

Sebagai permulaan,
gunakan method info() untuk memeriksa tipe data dan jumlah data yang ada
di dalam sales_df.

In [ ]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sales_id        5000 non-null   int64  
 1   order_id        5000 non-null   int64  
 2   product_id      5000 non-null   int64  
 3   price_per_unit  5000 non-null   int64  
 4   quantity        5000 non-null   int64  
 5   total_price     4981 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 234.5 KB


Berdasarkan hasil di atas, tidak ada masalah pada tipe data tiap kolom dalam
sales_df. Namun, terdapat keanehan pada jumlah data pada kolom
total_price.

In [ ]:
sales_df.isna().sum()

sales_id           0
order_id           0
product_id         0
price_per_unit     0
quantity           0
total_price       19
dtype: int64

Berdasarkan hasil di atas, dapat disimpulkan bahwa terdapat 19 missing
value pada kolom total_price.

In [ ]:
print("Jumlah duplicate: ", sales_df.duplicated().sum())
sales_df.describe()

Jumlah duplicate:  0


sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  4981.000000  
mean    206.307368  
std      86.352449  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

Hasil di atas menunjukkan bahwa tidak ada duplikasi pada sales_df. Selain
itu, ia juga menunjukkan tidak terdapat keanehan dalam ringkasan parameter
statistik dari sales_df.

Cleaning Data

Pembersihan atau cleaning data. Pada tahap ini, kita akan
membersihkan berbagai masalah yang telah teridentifikasi dalam proses
assessing data.

1. Membersihkan Data customer-df

  Berdasarkan hasil proses assessing data, diketahui bahwa terdapat tiga
masalah yang dijumpai dalam customer_df, yaitu duplicate data, missing
value, dan inaccurate value. Pada tahap ini, kita akan membersihkan ketiga
masalah tersebut.

a. Menghilangkan duplicate data

In [ ]:
customers_df.drop_duplicates(inplace=True)

In [ ]:
print("Jumlah duplikate: ", customers_df.duplicated().sum())

Jumlah duplikate:  0


b. Menangani missing value

  secara umum terdapat tiga metode untuk
  mengatasi missing value yaitu dropping, imputation, dan interpolation.
  Untuk menentukan metode mana yang akan digunakan, kita perlu
  melihat data yang mengandung missing value tersebut menggunakan
  teknik filtering seperti berikut.

In [ ]:
customers_df[customers_df.gender.isna()]

customer_id customer_name gender  age                     home_address  \
38            39      fulan 39    NaN   80  7440 Cameron Estate DrSuite 628   
167          168     fulan 168    NaN   27         2781 Berge MallSuite 452   
322          322     fulan 322    NaN   30        593 Becker CircleApt. 333   
393          393     fulan 393    NaN   34          5158 Levi HillSuite 531   
442          442     fulan 442    NaN   26           5157 Feil RoadApt. 633   
722          720     fulan 720    NaN   40         31 Jordan ParadeApt. 400   
745          743     fulan 743    NaN   57   09 Christopher StreetSuite 967   
773          771     fulan 771    NaN   74     7367 Wright JunctionApt. 773   
798          795     fulan 795    NaN   49          487 Summer MewsApt. 874   
801          798     fulan 798    NaN   56           27 Aiden KnollApt. 875   
825          822     fulan 822    NaN   59        41 Jenkins KnollSuite 438   
859          855     fulan 855    NaN   55       603 O'keefe KnollSuite 782   
863          859     fulan 859    NaN   38            32 Isla GroveApt. 078   
914          909     fulan 909    NaN   62    976 Murray Station StApt. 036   
934          929     fulan 929    NaN   68           394 Lily HillSuite 153   
948          943     fulan 943    NaN   64       3117 Heller PlaceSuite 149   
952          946     fulan 946    NaN   24      8227 Nicholas HillSuite 150   
994          988     fulan 988    NaN   35   1130 Turner Estate DrSuite 925   

     zip_code                   city                         state    country  
38       4622  North Victoriachester            Northern Territory  Australia  
167      1975         North Leoburgh             Western Australia  Australia  
322      1640             Jacobiview             Western Australia  Australia  
393      1474             Johnsburgh                    Queensland  Australia  
442      7249             Port Chloe               New South Wales  Australia  
722      1380             West Henry               South Australia  Australia  
745      6226         Lake Lukemouth             Western Australia  Australia  
773      8882               Kuhntown                      Victoria  Australia  
798      1712            East Hayden  Australian Capital Territory  Australia  
801      6531               Port Sam  Australian Capital Territory  Australia  
825      2588        Lake Andrewport               South Australia  Australia  
859      8822        Port Dylanmouth                      Tasmania  Australia  
863      7711            Rosechester               New South Wales  Australia  
914      3227               Langfort                      Tasmania  Australia  
934      2353             Beahanfurt            Northern Territory  Australia  
948       822           North Elijah               South Australia  Australia  
952       115           South Jasper                    Queensland  Australia  
994      9386              New Harry             Western Australia  Australia

Berdasarkan gambar di atas, dapat dilihat bahwa baris data tersebut masih
mengandung banyak informasi penting sehingga sayang jika langsung
dibuang. Oleh karena itu, pada kasus ini, kita akan menggunakan metode
imputation untuk menangani missing value.


Pada metode imputation, kita akan menggunakan nilai tertentu untuk
mengganti missing value. Musabab kolom gender merupakan kolom
kategorik, kita akan menggunakan nilai yang dominan sebagai pengganti
missing value tersebut. Gunakanlah method value_counts()untuk
mengidentifikasi nilai yang dominan.

In [ ]:
customers_df.gender.value_counts()

Prefer not to say    725
Male                 143
Female               115
Name: gender, dtype: int64

Berdasarkan hasil di atas, dapat diketahui bahwa nilai yang paling
dominan dalam kolom gender ialah “Prefer not to say”. Nilai inilah yang
selanjutnya akan kita gunakan sebagai pengganti missing value.
Proses penggantian ini dapat dilakukan menggunakan
method fillna()seperti contoh berikut.

In [ ]:
customers_df.fillna(value="Prefer not say", inplace=True)

In [ ]:
customers_df.isna().sum()

customer_id      0
customer_name    0
gender           0
age              0
home_address     0
zip_code         0
city             0
state            0
country          0
dtype: int64

c. Menangani inaccurate value

sekarang kita akan mengatasi masalah inaccurate value pada
kolom age. Sebagai awal, kita perlu melihat data baris data yang
mengandung inaccurate value tersebut (baris dengan nilai age
maksimum). Hal ini dilakukan menggunakan teknik filter seperti contoh
kode berikut.

In [ ]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
967          961     fulan 961  Prefer not to say  700   

                   home_address  zip_code        city            state  \
967  29 Farrell ParadeSuite 818      6528  New Joseph  South Australia   

       country  
967  Australia

Berdasarkan data tersebut, kita bisa berasumsi bahwa inaccurate value
tersebut terjadi karena human error sehingga kelebihan memasukkan
nilai nol. Oleh karena itu, gantilah dengan nilai 70. Proses ini dilakukan
dengan memanfaatkan method replace()seperti contoh berikut.

In [ ]:
customers_df.age.replace(customers_df.age.max(), 70, inplace=True)

In [ ]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
215          216     fulan 216  Prefer not to say  500   

               home_address  zip_code       city               state  \
215  038 Haley MewsApt. 810      3991  Bayertown  Northern Territory   

       country  
215  Australia

Penyebab kesalahan ini kemungkinan sama dengan sebelumnya,
yaitu human error yang kelebihan memasukkan nilai nol. Untuk
menangani hal ini, kita akan menggantinya dengan nilai 50.

In [ ]:
customers_df.age.replace(customers_df.age.max(), 50, inplace=True)

In [ ]:
customers_df.describe()

customer_id          age     zip_code
count  1001.000000  1001.000000  1001.000000
mean    500.942058    49.874126  5000.693307
std     289.013599    17.644663  2886.084454
min       1.000000    20.000000     2.000000
25%     251.000000    34.000000  2398.000000
50%     501.000000    50.000000  5079.000000
75%     751.000000    65.000000  7454.000000
max    1000.000000    80.000000  9998.000000

Berdasarkan hasil tersebut dapat dilihat bahwa kolom age memiliki nilai
maksimum yang cukup masuk akal. Selain itu, jika Ada perhatikan, nilai
mean dan standard deviation ikut berubah setelah kita menangani
inaccurate value tersebut.

2. Membersihkan Data orders_df

 Berdasarkan proses assessing data sebelumnya, diketahui bahwa
terdapat kesalahan tipe data untuk kolom order_date & delivery_date.

  Untuk mengatasi masalah ini, kita akan mengganti tipe data pada kolom
order_date & delivery_date menjadi datetime. Proses ini dapat dilakukan
menggunakan function to_datetime() yang disediakan oleh library pandas.

In [ ]:
datetime_columns = ["order_date", "delivery_date"]

for column in datetime_columns:
  orders_df[column] = pd.to_datetime(orders_df[column])

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   order_id       1000 non-null   int64         
 1   customer_id    1000 non-null   int64         
 2   payment        1000 non-null   int64         
 3   order_date     1000 non-null   datetime64[ns]
 4   delivery_date  1000 non-null   datetime64[ns]
dtypes: datetime64[ns](2), int64(3)
memory usage: 39.2 KB


3. Membersihkan Data product_df

Sesuai hasil
assessing data sebelumnya, kita mengetahui bahwa terdapat 6 data yang
terduplikasi pada product_df. Untuk mengatasi hal ini, kita perlu membuang
data yang sama tersebut menggunakan method drop_duplicates()

In [ ]:
product_df.drop_duplicates(inplace=True)

In [ ]:
print("Jumlah Duplikasi: ", product_df.duplicated().sum())

Jumlah Duplikasi:  0


Jika proses penghapusan data yang duplikat berjalan dengan lancar, kode di
atas akan menghasilkan keluaran seperti berikut “Jumlah duplikasi:  0”.

4. Membersihkan Data sales_df

Berdasarkan penilaian sebelumnya, diketahui bahwa terdapat 19 missing value pada kolom total_price. untuk mengetahui penanganan missing value yang paling sesuai, kita perlu melihat terlebih dahulu baris data yang mengandung missing valur tersebut.

In [ ]:
sales_df[sales_df.total_price.isna()]

sales_id  order_id  product_id  price_per_unit  quantity  total_price
9            9         2        1196             105         1          NaN
121        121        27        1027              90         3          NaN
278        278        63         360              94         2          NaN
421        421        95        1091             115         1          NaN
489        489       108        1193             105         3          NaN
539        539       117         405             119         2          NaN
636        636       134         653              93         3          NaN
687        687       145        1138             102         1          NaN
854        854       177          64             104         1          NaN
1079      1079       222         908              94         3          NaN
1193      1193       248        1121             102         2          NaN
1313      1313       272         826             117         1          NaN
1548      1548       316         103             118         3          NaN
1688      1688       345         428             107         1          NaN
1775      1775       359         694             113         2          NaN
1902      1902       381        1218             105         3          NaN
2025      2025       408         611             112         3          NaN
2164      2164       436         583             100         3          NaN
2347      2347       476         696             113         2          NaN

Berdasarkan tampilan data tersebut, kita menemukan bahwa nilai total_price
merupakan hasil perkalian antara price_per_unit dan quantity. Kita dapat
menggunakan pola ini untuk menangani missing value pada kolom
total_price.

In [ ]:
sales_df["total_price"] = sales_df["price_per_unit"] * sales_df["quantity"]

Kode di atas akan mengatasi seluruh missing value serta memastikan nilai
pada kolom total_price telah sesuai. Untuk memastikan hal ini, Anda dapat
memeriksa kembali jumlah missing value pada sales_df menggunakan kode
berikut.

In [ ]:
sales_df.isna().sum()

sales_id          0
order_id          0
product_id        0
price_per_unit    0
quantity          0
total_price       0
dtype: int64

**Latihan Exploratory Data Analysis**

**A. Menentukan Pertanyaan Bisnis**

Pada tahap ini, kita akan mendefinisikan berbagai
pertanyaan bisnis yang akan digunakan sebagai kompas dalam proses
eksplorasi data.


Berdasarkan beberapa masalah tersebut, kita dapat mendefinisikan
beberapa pertanyaan bisnis seperti berikut.

a. Bagaimana performa penjualan dan revenue perusahaan dalam
beberapa bulan terakhir?

b. Produk apa yang paling banyak dan paling sedikit terjual?

c. Bagaimana demografi pelanggan yang kita miliki?
Berdasarkan beberapa masalah tersebut, kita dapat mendefinisikan
beberapa pertanyaan bisnis seperti berikut.

d. Bagaimana performa penjualan dan revenue perusahaan dalam
beberapa bulan terakhir?

e. Produk apa yang paling banyak dan paling sedikit terjual?

f. Bagaimana demografi pelanggan yang kita miliki?



**B. Melakukan Eksplorasi Data**

Pada latihan ini, kita akan melakukan eksplorasi data guna menjawab
keenam pertanyaan tersebut. Pada tahap ini, mungkin hasil yang diperoleh
belum cukup untuk menjawab tersebut secara menyeluruh. Namun, proses
ini akan sangat membantu kita untuk mengenal data yang sedang ditangani
dan membuat strategi analisis yang paling efisien.

**1. Eksplorasi Data customers_df**

In [ ]:
customers_df.describe(include="all")

customer_id customer_name             gender          age  \
count   1001.000000          1001               1001  1001.000000   
unique          NaN          1000                  4          NaN   
top             NaN     fulan 943  Prefer not to say          NaN   
freq            NaN             2                725          NaN   
mean     500.942058           NaN                NaN    49.874126   
std      289.013599           NaN                NaN    17.644663   
min        1.000000           NaN                NaN    20.000000   
25%      251.000000           NaN                NaN    34.000000   
50%      501.000000           NaN                NaN    50.000000   
75%      751.000000           NaN                NaN    65.000000   
max     1000.000000           NaN                NaN    80.000000   

                      home_address     zip_code     city            state  \
count                         1001  1001.000000     1001             1001   
unique                        1000          NaN      961                8   
top     3117 Heller PlaceSuite 149          NaN  New Ava  South Australia   
freq                             2          NaN        3              140   
mean                           NaN  5000.693307      NaN              NaN   
std                            NaN  2886.084454      NaN              NaN   
min                            NaN     2.000000      NaN              NaN   
25%                            NaN  2398.000000      NaN              NaN   
50%                            NaN  5079.000000      NaN              NaN   
75%                            NaN  7454.000000      NaN              NaN   
max                            NaN  9998.000000      NaN              NaN   

          country  
count        1001  
unique          1  
top     Australia  
freq         1001  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

Berikut merupakan rangkuman parameter statistik yang diperoleh dari kode
di atas.
Berdasarkan rangkuman parameter statistik di atas, kita akan memperoleh
informasi jumlah pelanggan sebanyak 1007 orang yang berumur di kisaran
antara 20 hingga 80 tahun dengan rata-rata umur sebesar 49.87 tahun
dengan standar deviasinya sebesar 17.64 tahun. Informasi ini tentunya dapat memberikan gambaran bahwa segmentasi pelanggan yang kita miliki cukup
luas mulai dari remaja hingga lanjut usia.   


Sekarang coba kita lihat demografi pelanggan berdasarkan jenis kelamin
(gender). Untuk melakukan ini, kita akan menggunakan
method groupby() yang diikuti dengan method agg(). Berikut contoh kode
untuk melakukannya.

In [ ]:
customers_df.groupby(by="gender").agg({
    "customer_id" : "nunique",
    "age" : ["max", "min", "mean", "std"]
})

customer_id age                          
                      nunique max min       mean        std
gender                                                     
Female                    115  79  20  49.147826  16.646607
Male                      143  80  20  51.230769  18.462635
Prefer not say             18  80  24  48.777778  17.508168
Prefer not to say         725  80  20  49.748966  17.658998

Kode di atas akan menghasilkan sebuah pivot table. Ia memuat informasi
jumlah pelanggan (nunique dari customer_id) serta parameter umur yang
dikelompokkan berdasarkan jenis kelamin. Gambar di bawah ini merupakan
tampilan dari pivot table tersebut.

Berdasarkan pivot table di atas, dapat diketahui bahwa pelanggan yang kita
miliki didominasi oleh jenis kelamin prefer not to say. Di lain sisi, distribusi
umur mereka ternyata cukup mirip yaitu berkisar antara 20 hingga 80 tahun.

Selanjutnya, kita coba melihat persebaran jumlah pelanggan berdasarkan
kota (city) dan negara bagian (state). Untuk melakukannya kita juga akan
menggunakan method groupby(). Selain itu, agar hasilnya lebih mudah untuk
dilihat, kita akan mengurutkan nilainya menggunakan
method sort_values() secara descending.

In [ ]:
customers_df.groupby(by="city").customer_id.nunique().sort_values(ascending=False)

city
East Aidan       3
East Sophia      3
New Ava          3
West Jackfort    2
Corkeryshire     2
                ..
Jordanside       1
Josephland       1
Josephmouth      1
Justinport       1
Zacville         1
Name: customer_id, Length: 961, dtype: int64

In [ ]:
customers_df.groupby(by="state").customer_id.nunique().sort_values(ascending=False)

state
South Australia                 139
Queensland                      134
New South Wales                 132
Northern Territory              125
Western Australia               124
Australian Capital Territory    121
Victoria                        121
Tasmania                        104
Name: customer_id, dtype: int64

Berdasarkan hasil di atas, dapat dilihat bahwa persebaran pelanggan kita
cukup merata pada setiap kota dan negara bagian. Pelanggan kita paling
banyak berada di kota East Aidan, East Sophia, dan New Ava dengan jumlah
masing-masing tiga pelanggan. Selain itu, pelanggan kita paling banyak
berasal dari negara bagian South Australia.
Oke, sekarang kita telah memiliki cukup banyak informasi terkait data
customers_df.

**2. Eksplorasi Data orders_df**

Data kedua yang akan kita eksplor adalah orders_df. Ia mengandung
berbagai informasi terkait sebuah order yang terdiri dari order_id,
customer_id, order_date, dan delivery_date. Berdasarkan informasi tersebut,
kita bisa membuat sebuah kolom baru untuk memuat waktu pengiriman tiap
order. Untuk melakukannya, kita perlu menghitung selisih
antar delivery_date dan order_date serta menyimpannya
sebagai delivery_time. Selanjutnya, kita akan menggunakan
method apply() untuk melakukan sebuah operasi terhadap setiap elemen
dalam sebuah kolom DataFrame atau Series (bentuk satu dimensi dari
DataFrame). Operasi yang akan kita lakukan ialah menghitung jumlah detik
dari delivery_time menggunakan method total_seconds(). Nilai tersebut
selanjutnya diubah ke dalam satuan hari (dibagi 86400) dan diambil bilangan
bulatnya saja.

In [ ]:
delivery_time = orders_df["delivery_date"] - orders_df["order_date"]
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
